sdk/nodejs/appservice/zMixins.ts

```diff
@@ -334,87 +444,123 @@ export class CallbackFunctionApp<C extends Context<R>, E, R extends Result> exte
             accountTier: "Standard",
             accountReplicationType: "LRS",
-        }, opts);
+        }, { parent: this });
 
         const container = args.container || new storageMod.Container(makeSafeStorageContainerName(name), {
             resourceGroupName: args.resourceGroupName,
             storageAccountName: account.name,
             containerAccessType: "private",
-        }, opts);
+        }, { parent: this });
+
+        const archive = args.archive !== undefined 
+            ? args.archive
+            : pulumi.output(produceDeploymentPackage(args)).apply(m => new pulumi.asset.AssetArchive(m));
 
-        const assetMap = serializeCallback(name, args, bindings);
         const zipBlob = new storageMod.ZipBlob(name, {
             resourceGroupName: args.resourceGroupName,
             storageAccountName: account.name,
             storageContainerName: container.name,
             type: "block",
-            content: assetMap.apply(m => new pulumi.asset.AssetArchive(m)),
-        }, opts);
+            content: archive,
+        }, { parent: this });
 
         const codeBlobUrl = storageMod.signedBlobReadUrl(zipBlob, account);
-        super(name, {
+        this.app = new FunctionApp(name, {
             ...args,
-            ...resourceGroupArgs,
 
             version: "~2",
             appServicePlanId: plan.id,
             storageConnectionString: account.primaryConnectionString,
 
-            appSettings: pulumi.output(args.appSettings).apply(settings => {
-                settings = settings || {};
+            appSettings: combineAppSettings(args).apply(settings => {
                 return {
                     ...settings,
-                    WEBSITE_RUN_FROM_ZIP: codeBlobUrl,
+                    WEBSITE_RUN_FROM_PACKAGE: codeBlobUrl,
                     WEBSITE_NODE_DEFAULT_VERSION: util.ifUndefined(args.nodeVersion, "8.11.1"),
                 };
             }),
-        }, opts);
+        }, { parent: this });
 
         this.account = account;
         this.container = container;
         this.plan = plan;
         this.zipBlob = zipBlob;
+
+        const routePrefix = args.hostSettings 
+            && args.hostSettings.extensions 
+            && args.hostSettings.extensions.http 
+            && args.hostSettings.extensions.http.routePrefix;
+        const rootPath = routePrefix === "" ? "" : `${routePrefix === undefined ? "api" : routePrefix}/`;
+
+        this.endpoints = this.app.defaultHostname.apply(hostName => {
+            const endpoints = <{ [functionName: string]: string }>{};
+            if (args.functions !== undefined) {
+                const functionsWithEndpoints =  args.functions.filter(f => f.route !== undefined);
+                for (const f of functionsWithEndpoints) {
+                    endpoints[f.name] = `https://${hostName}/${rootPath}${f.route}`;
+                }
+            } else {
+                endpoints[0] = `https://${hostName}/${rootPath}`;

```

Comment:
this seems a bit wonky... would someone have to know to look this up?  should we instead have a defaultEndpoint for the case of an archive, and where the first func endpoint would go?

Reply:
Agreed, I'm not particularly happy about these endpoints...
Options:
1. Always return just 1 root endpoint and leave routing to users. A bit rudimentary: people might be used to seeing all URLs in Functions CLI / Visual Studio tools.
2. Somehow propagate URLs back to function objects? How? What if a function is in two apps?
3. Use a more sophisticated structure instead of key-value map.
4. ?

Reply:
1. i don't mind having the mapping object like you do.  but i would only do it for named key/value mappings.  I wouldn't have the `0` mapping.

You could also consider adding a `getEndpoint(data: Archive | Function)` which would the allow the user to pass in the same object they created the functionapp from to get the endpoint.  thougts?

Reply:
`getEndpoint` makes sense. Maybe just `getEndpoint(Function?)` - undefined giving you root?

Reply:
that sounds great!